[#9] 테스트코드 리팩토링

codesejin opened the pull request:
- 관련 Issue : https://github.com/f-lab-edu/off-coupon/issues/9

# Jacoco 보고서
- 테스트 코드 커버리지
  - 테스트된 영역 92 %
  - 조건부 분기 79 %
![image](https://github.com/f-lab-edu/off-coupon/assets/101460733/1806ffae-30fd-4a44-b0d1-bddfbbb5e7a8)

codesejin committed:
docs : README.md 수정

codesejin committed:
docs : Update README.md

f-lab-bot committed:
자동 PR 리뷰를 위한 workflow 파일 추가 by f-lab

codesejin committed:
chore : "JPA"에서 "Mybatis"로 변경

Subquery, Join 등 코드 리뷰를 더 세심하게 받기 위함

codesejin committed:
Merge pull request from master

codesejin committed:
docs : Read Me ERD 등 추가

codesejin committed:
feat : Mybatis config 설정 및 테스트

codesejin committed:
feat : 회원 가입시 유저 정보 유효성 검사

codesejin committed:
test : 유효성 검사 단위 테스트

codesejin committed:
feat : 회원가입 기능 구현

codesejin committed:
feat : Global ExceptionHandler 추가

codesejin committed:
feat : Spring security 추가, Bcrypt클래스 사용

codesejin committed:
test : Controller 회원가입 시 유효성 검증 테스트

codesejin committed:
feat : 이메일 중복 확인 및 서비스 레이어 테스트

codesejin committed:
test : member repository 회원가입 테스트

codesejin committed:
chore : jacoco 설정 변경
exception, util, OffcouonApplication.class 커버리지 제외

codesejin committed:
refactor : 변수명 수정 -> SonarQube에서 하드코딩이라도 판단
https://sonarcloud.io/project/security_hotspots?id=f-lab-edu_off-coupon&pullRequest=2&hotspots=AY1UfgtGw-sTzdJ5lVA-

codesejin committed:
docs : ERD 다이어그램 수정

codesejin committed:
style : Security config 람다표현식 메서드 레퍼런스로 정리

codesejin committed:
feat : 비밀번호 검증 커스텀 어노테이션

jakarta.validation.constraints 패키지의 @Pattern을 사용할 경우, 정규식이 점점 더 복잡해지면서, 유지 보수가 어려움.
요청에 있는 모든 필드를  Handler Exception Resolver에서 같이 검증하기 위해 별도의 커스텀 어노테이션 추가

codesejin committed:
Merge branch 'feature/signup' of https://github.com/f-lab-edu/off-coupon into feature/signup

codesejin committed:
refactor : password 어노테이션에 isBlank로직 추가

@Password 어노테이션과 @NotBlank 어노테이션을 같이 사용할 경우, 어떤 어노테이션이 먼저 실행될지 모른다.
그래서 @Password 어노테이션 내부 로직에서 null 객체에게 메소드 호출할경우 NPE발생

codesejin committed:
refactor : primitive type 에 @Notnull 제거

codesejin committed:
fix : Password 어노테이션 null 체크 로직 수정

codesejin committed:
refactor : ErrorMessage 클래스명 변경 및 매직 레터럴 추가

codesejin committed:
test : MemberControllerTest 회원가입 시 모든 필드에 대한 Null 체크 테스트

codesejin committed:
refactor : Member의 필드  어노테이션 변경

codesejin committed:
refactor : 비밀번호 암호화 책임 변경

기존 : 도메인 클래스에서 비밀번호 암호화 로직이 포함되어있었음
문제 : 비밀번호 암호화는 비즈니스 로직
변경 : 서비스 레이어에게 암호화 로직 책임 할당

codesejin committed:
refactor : CustomException 클래스 명 변경

추후 익셉션 종류별로 만들게 되면 CustomXXXException 로 만들어야 하기 때문에 유지보수성을 위해 수정

codesejin committed:
refactor : 중복 이메일 검증 쿼리 수정

count는 모든 열을 조회하기 때문에, 해당 이메일이랑 일치하는 유저가 존재하는지 확인하는 exists로 변경

codesejin committed:
refactor : BCryptPasswordEncryptor로 Util성 클래스명에 맞도록 수정

codesejin committed:
chore : properties local 설정 분리

codesejin committed:
chore : gitignore, properties분리, 트랜잭션 어노테이션 추가

codesejin committed:
refactor : DATETIME에서 TIMESTAMP로 변경

codesejin committed:
refactor : SonarCloud 관련 이슈 리팩토링

codesejin committed:
refactor : 2차 리뷰 리팩토링

1. 유틸성 클래스 유효성 검사 추가
2. 필요없는 변수 제거, 메소드 반환값 바로 이용하기
3. 논리 조건문 가독성 좋게 수정
4. else if문 대신 Stream API의 anyMatch() 활용

codesejin committed:
refactor : 불변 객체로 변경

codesejin committed:
test : 변경된 로직과 테스트코드 일관되게 수정

codesejin committed:
test : 필요없는 메소드 제거

codesejin committed:
refactor : 필요없는 변수  제거, early 리턴

codesejin committed:
docs : ERD 다이어그램 수정

codesejin committed:
refactor : ResponseDTO의 data 제네릭으로 변경

codesejin committed:
Merge branch 'feature/signup' of https://github.com/f-lab-edu/off-coupon into feature/signup

codesejin committed:
feat : 쿠폰 발행 이벤트 엔티티 작성

codesejin committed:
test: Event, Coupon 도메인 요구사항 검증 테스트

기간 검증, 시간 검증, 잔여 수량 검증

codesejin committed:
chore : merge from develop

codesejin committed:
feat : 쿠폰, 이벤트 도메인 익셉션

codesejin committed:
feat : 쿠폰 발급 기능 서비스 레이어

codesejin committed:
test : 쿠폰 발급 기능 동시성 테스트

Coupon 테이블의 issuedQuantity가 의도대로 반영 되지 않음을 확인

codesejin committed:
fix : 익셉션 응답에 추가, 중복발급 로직 수정

codesejin committed:
refactor : HttpStatus, 중복 문자열 매직리터럴로 변경

codesejin committed:
refactor : 경로 변경 및 디렉토리 변경

codesejin committed:
test : synchronized 동시성 테스트

codesejin committed:
test : 비관적락, 낙관적락 동시성 테스트

codesejin committed:
test : NamedLock 동시성 테스트

codesejin committed:
test : Redisson pub/sub 동시성테스트

codesejin committed:
style : 사용하지 않는 import 지우기

codesejin committed:
feature : 이벤트 조회 Cache 적용

codesejin committed:
refactor : 다양한 락 기법 리팩토링 및 비동기 버전 추가

codesejin committed:
style : 주석에  코드 설명 추가

codesejin committed:
style : 사용하지 않는 import제거

codesejin committed:
style : 주석 리팩토링

codesejin committed:
style : 주석 리팩토링

codesejin committed:
refactor : 하드코딩 -> 매직넘버, Duration, Period 클래스 사용

codesejin committed:
refactor : 트랜잭션 Proxy객체에서 대상 객체 내부 메소드 호출 시 트랜잭션 적용안되는 것 해결

codesejin committed:
refactor : 중복코드 제거

codesejin committed:
refactor : private 메소드로 접근제어자 변경

codesejin committed:
refactor : 패키지명에 대분자 사용 x, 언더바로 변경

codesejin committed:
refactor : 제네릭 파라미터 추가

codesejin committed:
refactor : 테스트코드에 public제거, Redis key  매직레터럴 사용

codesejin committed:
refactor : 단일 statement일 경우 중괄호 제거

codesejin committed:
refactor : Boolean을 조건절에 사용할 경우 null 체크

codesejin committed:
refactor : 제네릭 파라미터 추가

codesejin committed:
refactor : Lettuce, Redisson client 동시 사용 가능하도록 수정

codesejin committed:
test : 쿠폰 발행 테스트 코드 추가

sonarcloud에서 코드 커버리지 충족 못 시키면 A레벨을 안주는 것 같아요

codesejin committed:
refactor : 생성자 파라미터 줄이기

codesejin committed:
refactor : record파일 수정 및 InterruptedException 예외 전환 처리

record 파일은 모든 멤버변수를 인자로 하는 public 생성자를 자동으로 생성한다

codesejin committed:
refactor : 쿠폰, 이벤트 엔티티 테스트코드 객체 생성 인자 변경

codesejin committed:
chore : CICD 연습용 properties 수정

codesejin committed:
refactor : CICD 빌드시 테스트실패하는거 수정

codesejin committed:
test : CICD 빌드시 테스트 실패 - CouponIssueRequestService 테스트 추가

codesejin committed:
chore : Jacoco 코드 커버리지 임시 수정 - CICD 테스트중

codesejin committed:
chore : Jacoco 코드 커버리지 범위 임시  수정 - CICD 테스트중

codesejin committed:
chore : Dockerfile 테스트

codesejin committed:
chore : CICD 테스트 - Redis Port 타입변환

codesejin committed:
chore : CICD 테스트 - Redis Port 타입변환

codesejin committed:
refactor : 접근제어자 private으로 캡슐화

codesejin committed:
test : cicd 테스트

codesejin committed:
test : 추상화하고 재사용 가능한 유틸 테스트클래스

codesejin committed:
test : 충돌 해결

codesejin committed:
test : CouponIssueController 테스트

codesejin committed:
test : 이벤트 기간, 시간 null 체크 테스트코드 추가

codesejin committed:
test : 분산락 컴포넌트클래스 테스트

codesejin committed:
refactor : 소나클라우드 코드 개선

codesejin committed:
chore : 코드 커버리지 범위 변경

f-lab-maverick commented on src/test/java/com/flab/offcoupon/service/coupon_issue/sync/RedissonLockCouponIssueTest.java:
```diff
@@ -37,42 +32,20 @@ class RedissonLockCouponIssueTest {
-                LocalDateTime.now());
-        couponRepository.save(coupon);
+    void setUp() {
+        setupUtils.setUpEventAndCoupon(eventRepository, couponRepository);
```
이렇게 `set~` 메서드를 통해 무조건 초기화를 하고 사용해야 하는 경우, 1. 객체생성 -> 2. 초기화로 과정이 두개의 단계를 가지게 되고, 경우에 따라서는 과정이 누락되거나 초기화 전 객체의 사용이 발생될 수 있습니다.

그런 경우가 발생하지 않도록 아예 객체 생성시 꼭 필요한 상태를 생성자로 주입해주면 객체 생성부터 세팅까지를 한단계로 줄일 수 있으니 생성자를 통해 상태를 주입해주세요~

각 테스트마다 독립된 셋업이라면 이렇게 만들어줄 수 있습니다.


```suggestion
        setupUtils  =new SetupUtils(eventRepository, couponRepository);
```

codesejin replied:
초기화 메소드 내부에서 생성자 주입하는게 휴먼 에러를 방지하고 훨씬 유지보수에 쉬운거군요 감사합니당~

f-lab-maverick reviewed (approved):
중복되는 부분을 모두 공통화하고, 테스트코드를 깔끔하게 잘 만드셨네요~
고생하셨습니다. 머지해주셔도 좋습니다.

codesejin committed:
refactor : 객체 생성과 setup(초기화)를 한 단계로 줄이기
1. 객체 생성 -> 2. 초기화 로 2단계를 가지게 되면
경우에 따라서 가정이 누락되거나 초기화 전에 객체 사용이 발생할 수 있음
그런 경우를 예방하기 위해 아예 객체 생성 시 꼭 필요한 상태를 생성자로 주입하면 객체 생성부터 초기화까지 1단계로 줄일 수 있음
